### ISSUE TITLE
[12.0] file base/data/res.lang.csv could not be processed

### ISSUE THREAD

**Issue by rvalyi** [OCA/OpenUpgrade#1665]
Impacted versions: 12.0

Steps to reproduce:
Try to migrate a production DB with the base module installed

Current behavior:
Halt with:
```
2018-12-14 16:20:28,333 68 INFO openupgrade-to-12 odoo.models: Actual recompute of field res.currency.decimal_places for 176 recs.                                                                          
2018-12-14 16:20:28,764 68 INFO openupgrade-to-12 odoo.modules.loading: loading base/data/res.lang.csv                                                                                                       
2018-12-14 16:20:30,339 68 WARNING openupgrade-to-12 odoo.modules.loading: Transient module states were reset                                                                               
2018-12-14 16:20:30,340 68 ERROR openupgrade-to-12 odoo.modules.registry: Failed to load registry                                                                                    
Traceback (most recent call last):                                                                                                                                                                           
  File "/odoo/src/odoo/modules/registry.py", line 86, in new                                                                                                                                                 
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)                                                                                                     
  File "/odoo/src/odoo/modules/loading.py", line 414, in load_modules                                                                                                                           
    report=report, models_to_check=models_to_check, upg_registry=upg_registry)                                                                                                                               
  File "/odoo/src/odoo/modules/loading.py", line 251, in load_module_graph                                                                                                                                   
    load_data(cr, idref, mode, kind='data', package=package, report=report)                                                                                                               
  File "/odoo/src/odoo/modules/loading.py", line 72, in load_data                                                                                                                                            
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind, report)                                                                                                                      
  File "/odoo/src/odoo/tools/convert.py", line 802, in convert_file                                                                                                                                          
    convert_csv_import(cr, module, pathname, fp.read(), idref, mode, noupdate)                                                                                                                              
  File "/odoo/src/odoo/tools/convert.py", line 849, in convert_csv_import                                                                                                               
    raise Exception(_('Module loading %s failed: file %s could not be processed:\n %s') % (module, fname, warning_msg))                                                                                      
Exception: Module loading base failed: file base/data/res.lang.csv could not be processed:                                                                                                                   
 Erreur inconnue pendant l'importation : <class 'odoo.exceptions.UserError'>: ('Le code langue ne peut pas être modifié.', '')                                                                               
2018-12-14 16:20:30,342 68 CRITICAL openupgrade-to-12 odoo.service.server: Failed to initialize database `openupgrade-to-12`.                                                                            
Traceback (most recent call last):                                                                                                      
  File "/odoo/src/odoo/service/server.py", line 1042, in preload_registries                                                                                                                                  
    registry = Registry.new(dbname, update_module=update_module)                                                                                                                             
  File "/odoo/src/odoo/modules/registry.py", line 86, in new                                                                                                     
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)                                                                                                                               
  File "/odoo/src/odoo/modules/loading.py", line 414, in load_modules                                                                                                                                        
    report=report, models_to_check=models_to_check, upg_registry=upg_registry)                                                                                                          
  File "/odoo/src/odoo/modules/loading.py", line 251, in load_module_graph                                                                                                                                   
    load_data(cr, idref, mode, kind='data', package=package, report=report)                                                                                                                                  
  File "/odoo/src/odoo/modules/loading.py", line 72, in load_data                                                                                       
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind, report)                                                                                                               
  File "/odoo/src/odoo/tools/convert.py", line 802, in convert_file                                                                                                           
    convert_csv_import(cr, module, pathname, fp.read(), idref, mode, noupdate)                                                                                                                               
  File "/odoo/src/odoo/tools/convert.py", line 849, in convert_csv_import                                                                                                                                    
    raise Exception(_('Module loading %s failed: file %s could not be processed:\n %s') % (module, fname, warning_msg))                    
Exception: Module loading base failed: file base/data/res.lang.csv could not be processed:                                                                                                                  
 Erreur inconnue pendant l'importation : <class 'odoo.exceptions.UserError'>: ('Le code langue ne peut pas être modifié.', '')                                                                               
2018-12-14 16:20:30,342 68 INFO openupgrade-to-12 odoo.service.server: Stopping gracefully  
```

Expected behavior:

migrate the base module

More explanation:

I don't know yet if there is something we should do in the v12 migration of the base module and/or if it is because something is corrupted in the database I'm trying to migrate. This is a production database that was used without problem on v8 and migrated to v9, v10 and v11 without problem apparently.

Workaround:

Apparently, commenting the ```base/data/res.lang.csv``` entry in the __manifest__.py makes the script continues further to the next modules...

I'm not telling this is a bug in OpenUpgrade, I'm only logging the issue so that eventually others with the same issue could team on that.

**Comment by pedrobaeza** [comment]
Here it's the answer :wink: :

https://github.com/odoo/odoo/commit/40b879b75dfe271558e8ea05a3911b5be1ee80d6#diff-ca8c96634039bcd7e57220e36ca8eda1

The curious thing is that this doesn't happen in current test DB.

**Comment by MiquelRForgeFlow** [comment]
@rvalyi the good fix should be to manually fix the Khmer code in your db.

**Comment by pedrobaeza** [comment]
But we should do the same in OpenUpgrade `base` if this is not possible to be handled by regular update.

**Comment by wamdam** [comment]
hit me too

**Comment by linuxnow** [comment]
2025-06-27 00:29:03,749 1 ERROR metropol odoo.modules.registry: Failed to load registry 
Traceback (most recent call last):                                    
  File "/opt/odoo/custom/src/odoo/odoo/modules/registry.py", line 87, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/custom/src/odoo/odoo/modules/loading.py", line 369, in load_modules
    report=report, models_to_check=models_to_check)
  File "/opt/odoo/custom/src/odoo/odoo/modules/loading.py", line 225, in load_module_graph
    load_data(cr, idref, mode, kind='data', package=package, report=report)
  File "/opt/odoo/custom/src/odoo/odoo/modules/loading.py", line 68, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind, report)
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 732, in convert_file
    convert_csv_import(cr, module, pathname, fp.read(), idref, mode, noupdate)
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 779, in convert_csv_import
    raise Exception(_('Module loading %s failed: file %s could not be processed:\n %s') % (module, fname, warning_msg))
Exception: Module loading base failed: file base/data/res.lang.csv could not be processed:
 Error desconegut durant la importació: <class 'odoo.exceptions.UserError'>: ("El codi d'idioma no pot ser modificat", '')
Error desconegut durant la importació: <class 'odoo.exceptions.UserError'>: ("El codi d'idioma no pot ser modificat", '')
Error desconegut durant la importació: <class 'odoo.exceptions.UserError'>: ("El codi d'idioma no pot ser modificat", '')
2025-06-27 00:29:03,750 1 CRITICAL metropol odoo.service.server: Failed to initialize database `metropol`. 
Traceback (most recent call last):                                    
  File "/opt/odoo/custom/src/odoo/odoo/service/server.py", line 1194, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)      
  File "/opt/odoo/custom/src/odoo/odoo/modules/registry.py", line 87, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/opt/odoo/custom/src/odoo/odoo/modules/loading.py", line 369, in load_modules
    report=report, models_to_check=models_to_check)                   
  File "/opt/odoo/custom/src/odoo/odoo/modules/loading.py", line 225, in load_module_graph
    load_data(cr, idref, mode, kind='data', package=package, report=report)
  File "/opt/odoo/custom/src/odoo/odoo/modules/loading.py", line 68, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind, report)
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 732, in convert_file
    convert_csv_import(cr, module, pathname, fp.read(), idref, mode, noupdate)
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 779, in convert_csv_import
    raise Exception(_('Module loading %s failed: file %s could not be processed:\n %s') % (module, fname, warning_msg))
Exception: Module loading base failed: file base/data/res.lang.csv could not be processed:
 Error desconegut durant la importació: <class 'odoo.exceptions.UserError'>: ("El codi d'idioma no pot ser modificat", '')
Error desconegut durant la importació: <class 'odoo.exceptions.UserError'>: ("El codi d'idioma no pot ser modificat", '')


I cannot migrate to version 13 using doodba because I can't solve this issue. I' have tried different potsgres versions, manually updating km in the db, all that I could think of, dump/restore.... no way.

Can somebody point me to a workaround?

**Comment by pedrobaeza** [comment]
@linuxnow check the languages codes, as they are modified in your DB, and you have to change them by SQL before migrating.

**Comment by linuxnow** [comment]
I see that's the problem but I can't figure out what to do. Can you tell me how to check the current language code and how to change it? I can use psql, no problem

**Comment by pedrobaeza** [comment]
You have to check each language code and the file to see which ones are different, and fix it via SQL.

**Comment by linuxnow** [comment]
I have no idea on how to do it, any help would be appreciated otherwise I'm stuck.
Please tell me how to compare it so that I can fix it, I'm not aware that I made any change.

**Comment by linuxnow** [comment]
My problem is that I don't know where are the new language codes to compare with the existing ones. I'm not used to using openugrade. If you please tell me how can I reach these new codes I'll be able to update the ones that have changed.

Thanks in advance Pedro :)

**Comment by pedrobaeza** [comment]
Check the file https://github.com/odoo/odoo/blob/12.0/odoo/addons/base/data/res.lang.csv

**Comment by linuxnow** [comment]
This is my current data:
id, name, code, iso_code, translatable, active, direction, date_format, time_format, "grouping", decimal_point, thousands_sep, create_uid, create_date, write_uid, write_date, week_start
16      English (AU)    en_AU   en_AU   t       f       ltr     %d/%m/%Y        %H:%M:%S        [3,0]   .       ,       1       2019-01-16 12:43:59.396458      1       2025-06-12 14:17:41.33033       7
17      English (UK)    en_GB   en_GB   t       f       ltr     %d/%m/%Y        %H:%M:%S        [3,0]   .       ,       1       2019-01-16 12:43:59.396458      1       2025-06-12 14:17:41.33033       7
18      Estonian / Eesti keel   et_EE   et      t       f       ltr     %d.%m.%Y        %H:%M:%S        [3,0]   ,               1       2019-01-16 12:43:59.396458      1       2025-06-12 14:17:41.33033       1
19      Finnish / Suomi fi_FI   fi      t       f       ltr     %d.%m.%Y        %H.%M.%S        [3,0]   ,               1       2019-01-16 12:43:59.396458      1       2025-06-12 14:17:41.33033       1

Everything is the same but the id field that has changed completely:
Now we have base.lang_en_AU instead of a numeric code.

This is the sequence:
CREATE SEQUENCE public.res_lang_id_seq
    AS integer
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1;
ALTER TABLE public.res_lang_id_seq OWNER TO odoo;
ALTER SEQUENCE public.res_lang_id_seq OWNED BY public.res_lang.id;
ALTER TABLE ONLY public.res_lang ALTER COLUMN id SET DEFAULT nextval('public.res_lang_id_seq'::regclass);

I'd be very grateful if you could confirm that this is correct.

Thank you Pedro

**Comment by pedrobaeza** [comment]
No, there are hundreds of records, not only 4. Please check all of them.

**Comment by linuxnow** [comment]
Sorry for the misunderstanding, I just copied a few records not all the 84.

I mean that this is the change that I have to do manually, update all the ids.

**Comment by pedrobaeza** [comment]
No, you have to avoid duplicated code. Maybe the problem is that one language record has lost its XML-ID, and now when it's recreated on load, it's duplicated.

**Comment by linuxnow** [comment]
I don't understand it.

My current ids are numeric, based on a sequence, from 1 to 84. Here are 2 records:

 **id**, name, code, iso_code, translatable, active, direction, date_format, time_format, "grouping", decimal_point, thousands_sep, create_uid, create_date, write_uid, write_date, week_start
**16** English (AU) en_AU en_AU t f ltr %d/%m/%Y %H:%M:%S [3,0] . , 1 2019-01-16 12:43:59.396458 1 2025-06-12 14:17:41.33033 7
**17** English (UK) en_GB en_GB t f ltr %d/%m/%Y %H:%M:%S [3,0] . , 1 2019-01-16 12:43:59.396458 1 2025-06-12 14:17:41.33033 7

The file odoo/addons/base/data/res.lang.csv  that you mentioned here
> Check the file https://github.com/odoo/odoo/blob/12.0/odoo/addons/base/data/res.lang.csv

has this data (only 2 records are copied):
"**id**","name","code","iso_code","direction","grouping","decimal_point","thousands_sep","date_format","time_format","translatable","week_start"
"**base.lang_en**","English","en_US","en","Left-to-Right","[3,0]",".",",","%m/%d/%Y","%H:%M:%S","True","7"
"**base.lang_am_ET**","Amharic / አምሃርኛ","am_ET","am_ET","Left-to-Right","[3,0]",".",",","%d/%m/%Y","%I:%M:%S","True","7"
...

I've checked all the codes and the only difference I can see id that it id filed has changed from a numeric sequence to a string (base.lang.XXXXX).

What I understand from your answer is that this is not what is causing the error. Please confirm it.

And I guess I have toi find a difference in the code field which I haven't found but I'll check again.

Thanks Pedro, I appreciate your help.

**Comment by pedrobaeza** [comment]
Sorry, I can't continue helping you, as there's a base knowledge that you need to have to perform these things, knowing concepts like XML-IDs, and this has overpassed the time I can spent helping community.

**Comment by linuxnow** [comment]
Thanks for your help Pedro, fully understandable, your contribution to the community is highly praised.

**Comment by linuxnow** [comment]
I found the culprits::
-ar, ar_AR
-fil, fil
I could not fix the changed code error but I could disable the updates.
update ir_model_data set noupdate = 't' where name IN ('lang_fil', 'lang_ar');

I know it's not optimal but I have been unable to get the languages accepted.

After this openupgrade run smoothly until another error is raised. I have searched thoroughly but don't know how to continue, any pointer will be appreciated.

ValueError: Field `image` does not exist

Context error:
Check `res.users.groups`
[view_id: 148, xml_id: base.user_groups_view, model: res.users, parent_id: 144]

Full log attached.

[error_image.txt](https://github.com/user-attachments/files/21046123/error_image.txt)
